core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java

```diff
@@ -74,10 +94,51 @@ synchronized void releaseTranslogGenView(long translogGen) {
     /**
      * returns the minimum translog generation that is still required by the system. Any generation below
      * the returned value may be safely deleted
+     *
+     * @param readers current translog readers
+     * @param writer  current translog writer
      */
-    synchronized long minTranslogGenRequired() {
-        long viewRefs = translogRefCounts.keySet().stream().reduce(Math::min).orElse(Long.MAX_VALUE);
-        return Math.min(viewRefs, minTranslogGenerationForRecovery);
+    synchronized long minTranslogGenRequired(List<TranslogReader> readers, TranslogWriter writer) {
+        long minByView = getMinTranslogGenRequiredByViews();
+        long minByAge = getMinTranslogGenByAge(readers, writer);
+        long minBySize = getMinTranslogGenBySize(readers, writer);
+        long minByAgeAndSize = Math.max(minByAge, minBySize);
+        return Math.min(minByAgeAndSize, Math.min(minByView, minTranslogGenerationForRecovery));
+    }
+
+    private long getMinTranslogGenBySize(List<TranslogReader> readers, TranslogWriter writer) {
```

Comment:
do you wanna make them pkg private then you can test them in isolation. I'd love to see that